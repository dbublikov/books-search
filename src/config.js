import axios from 'axios';

const API_KEY = 'AIzaSyB0Gf_r2La1jo5bMrpUzgVaQqHkxqS05-A';

const baseUrl = 'https://www.googleapis.com/books/v1/volumes';

const getBooks = `${baseUrl}?q=subject:art&orderBy=newest&maxResults=12&key=${API_KEY}`;

export const initLoad = async () => {
  const res = await axios.get(getBooks);
  // console.log(res)
  if (res.isAxiosError) {
    throw new Error('Can\'t fetch');
  }
  return res;
};

export const loadDetails = async (id) => {
  const res = await axios.get(`${baseUrl}/${id}`);
  if (res.isAxiosError) {
    throw new Error('Can\'t fetch');
  }
  return res;
};
