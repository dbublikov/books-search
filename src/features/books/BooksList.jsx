import styled from 'styled-components';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import { clearControls } from '../controls/controls-slice';
import { initLoad, loadMore, selectBooks, selectBooksInfo } from './books-slice';

import { Card } from '../../components/Card';
import { List } from '../../components/List';
import { LoadMoreButton } from '../../components/Buttons';


const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
`;

export const BooksList = () => {
  const books = useSelector(selectBooks);
  const { status, qty, pagination } = useSelector(selectBooksInfo);
  const { currIndex, pageTerm, step } = pagination;

  const dispatch = useDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    if (!qty) {
      dispatch(initLoad());
    }

    return () => {
      dispatch(clearControls());
    };
  }, [qty, dispatch]);

  return (
    <Wrapper>
      {
        status === 'received' && (
          <List>
            {books.map(({ id, volumeInfo }) => {
              return (
                <Card
                  key={id}
                  id={id}
                  onClick={() => navigate(`/book/${id}`)}
                  {...volumeInfo}
                />
              );
            })}
          </List>
        )
      }
      {pageTerm && (currIndex <= qty-step) && pagination.status !== 'loading'
        && <LoadMoreButton onClick={() => dispatch(loadMore())}>Load more</LoadMoreButton>}
    </Wrapper>
  );
};
