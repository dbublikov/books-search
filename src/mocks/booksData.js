export const booksData = {
  items: [
    {
      id: 'osZVEAAAQBAJ',
      volumeInfo: {
        title: 'Start Painting Now',
        subtitle: 'Discover Your Artistic Potential',
        authors: [
          'Emily Powell',
          'Sarah Moore',
        ],
      },
    },
    {
      id: 'd_E3EAAAQBAJ',
      volumeInfo: {
        title: 'What the Ermine Saw',
        subtitle: 'The Extraordinary Journey of Leonardo da Vinci\'s Most Mysterious Portrait',
        authors: [
          'Eden Collinsworth',
        ],
      },
    },
    {
      id: 'gwJLEAAAQBAJ',
      volumeInfo: {
        title: 'What Artists Wear',
        authors: [
          'Charlie Porter',
        ],
      },
    },
  ],
  totalItems: 6,
};

export const extraBooksData = {
  items: [
    {
      id: '6DBNEAAAQBAJ',
      volumeInfo: {
        title: 'Looking for Theophrastus',
        subtitle: 'Travels in Search of a Lost Philosopher',
        authors: [
          'Laura Beatty',
        ],
      },
    },
    {
      id: 'tYkqEAAAQBAJ',
      volumeInfo: {
        title: 'Portrait of a Thief',
        authors: [
          'Grace D. Li',
        ],
      },
    },
    {
      id: 'xsNEEAAAQBAJ',
      volumeInfo: {
        title: 'The Art of Cloth in Mughal India',
        authors: [
          'Sylvia Houghteling',
        ],
      },
    },
  ],
};

