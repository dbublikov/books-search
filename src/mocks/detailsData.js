export const detailsData = {
  id: '2upIEAAAQBAJ',
  volumeInfo: {
    authors: ['Hugh Eakin'],
    title: 'Picasso\'s War',
    subtitle: 'How Modern Art Came to America',
    categories: ['Art / History / Modern (late 19th Century to 1945)'],
    description: '<b><i>NEW YORK TIMES</i> EDITORS’ CHOICE • A riveting story of how dueling ambitions and the power of prodigy made America the cultural center of the world—and Picasso the most famous artist alive—in the shadow of World War II<br><br>“Fascinating, eloquent, wonderfully lucid, <i>Picasso’s War</i> will change whatever we thought we knew about modern art and its complicated reception on this side of the Atlantic.”—Francine Prose, author of <i>The Vixen</i></b>',
    pageCount: 480,
    publisher: 'Crown',
    publishedDate: '2022-07-12',
  },
};


