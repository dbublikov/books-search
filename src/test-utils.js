import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';

import { configureStore } from '@reduxjs/toolkit';
import { controlsReducer } from './features/controls/controls-slice';
import { detailsReducer } from './features/details/details-slice';
import { booksReducer } from './features/books/books-slice';

const reducer = {
  books: booksReducer,
  controls: controlsReducer,
  details: detailsReducer,
};

export function getStoreWithState(preloadedState) {
  return configureStore({ reducer, preloadedState });
}

export function renderWithContext(element, state) {
  const store = getStoreWithState(state);
  const utils = render(
    <Provider store={store}>
      <Router>{element}</Router>
    </Provider>
  );
  return { store, ...utils };
}

export const getPreloadedState = () => {
  const state = {
    books: {
      status: 'idle',
      error: null,
      totalItems: null,
      list: [],
      pagination: {
        status: 'idle',
        error: null,
        currIndex: 2,
        pageTerm: 'placeholder',
        step: 2,
      },
    },
  };
  return state;
};
