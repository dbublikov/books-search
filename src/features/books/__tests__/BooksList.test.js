import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithContext, getPreloadedState } from '../../../test-utils';
import { BooksList } from '../BooksList';
import { booksData, extraBooksData } from '../../../mocks/booksData';
import * as api from '../../../config';

const getBooksSpyInit = jest.spyOn(api, 'initLoad');
const getBooksSpyLoad = jest.spyOn(api, 'loadBooks');

describe('BooksList', () => {
  it('should list several books initially', async () => {
    getBooksSpyInit.mockResolvedValue({ data: booksData });
    const view = renderWithContext(<BooksList />);

    await waitFor(() => expect(getBooksSpyInit).toHaveBeenCalledTimes(1));
    const articles = screen.getAllByRole('article');
    expect(articles).toHaveLength(3);
    expect(view).toMatchSnapshot();
  });

  it('should load more books when button clicked', async () => {
    getBooksSpyInit.mockResolvedValue({ data: booksData });
    getBooksSpyLoad.mockResolvedValue({ data: extraBooksData });

    const state = getPreloadedState();
    const view = renderWithContext(<BooksList />, state);
    // const {debug, store} = renderWithContext(<BooksList />, state)
    // debug()
    // console.log(store.getState())

    const button = await screen.findByRole('button', { name: /Load more/i });
    expect(button).toBeInTheDocument();
    expect(view).toMatchSnapshot();

    userEvent.click(button);
    expect(screen.getByTestId('spinner')).toBeVisible();
    await waitFor(() => expect(getBooksSpyLoad).toHaveBeenCalledTimes(1));

    expect(screen.getAllByRole('article')).toHaveLength(6);
  });
});





