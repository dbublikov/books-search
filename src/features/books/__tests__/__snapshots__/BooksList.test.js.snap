// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BooksList should list several books initially 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-eDWCr editcY"
      >
        <section
          class="sc-gKPRtg ilzhkz"
        >
          <article
            class="sc-bcXHqe UJFxh"
          >
            <img
              alt="Start Painting Now"
              class="sc-gswNZR bUwCaT"
              src="https://books.google.com/books/publisher/content/images/frontcover/osZVEAAAQBAJ?fife=w270"
            />
            <div
              class="sc-dkrFOg hXwiuy"
            >
              <h5
                class="sc-eDvSVe broPkG"
              >
                N/A
              </h5>
              <h3
                class="sc-hLBbgP degNmy"
              >
                Start Painting Now
              </h3>
              <h5
                class="sc-jSUZER hEWtoK"
              >
                Emily Powell, Sarah Moore
              </h5>
            </div>
          </article>
          <article
            class="sc-bcXHqe UJFxh"
          >
            <img
              alt="What the Ermine Saw"
              class="sc-gswNZR bUwCaT"
              src="https://books.google.com/books/publisher/content/images/frontcover/d_E3EAAAQBAJ?fife=w270"
            />
            <div
              class="sc-dkrFOg hXwiuy"
            >
              <h5
                class="sc-eDvSVe broPkG"
              >
                N/A
              </h5>
              <h3
                class="sc-hLBbgP degNmy"
              >
                What the Ermine Saw
              </h3>
              <h5
                class="sc-jSUZER hEWtoK"
              >
                Eden Collinsworth
              </h5>
            </div>
          </article>
          <article
            class="sc-bcXHqe UJFxh"
          >
            <img
              alt="What Artists Wear"
              class="sc-gswNZR bUwCaT"
              src="https://books.google.com/books/publisher/content/images/frontcover/gwJLEAAAQBAJ?fife=w270"
            />
            <div
              class="sc-dkrFOg hXwiuy"
            >
              <h5
                class="sc-eDvSVe broPkG"
              >
                N/A
              </h5>
              <h3
                class="sc-hLBbgP degNmy"
              >
                What Artists Wear
              </h3>
              <h5
                class="sc-jSUZER hEWtoK"
              >
                Charlie Porter
              </h5>
            </div>
          </article>
        </section>
        
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-eDWCr editcY"
    >
      <section
        class="sc-gKPRtg ilzhkz"
      >
        <article
          class="sc-bcXHqe UJFxh"
        >
          <img
            alt="Start Painting Now"
            class="sc-gswNZR bUwCaT"
            src="https://books.google.com/books/publisher/content/images/frontcover/osZVEAAAQBAJ?fife=w270"
          />
          <div
            class="sc-dkrFOg hXwiuy"
          >
            <h5
              class="sc-eDvSVe broPkG"
            >
              N/A
            </h5>
            <h3
              class="sc-hLBbgP degNmy"
            >
              Start Painting Now
            </h3>
            <h5
              class="sc-jSUZER hEWtoK"
            >
              Emily Powell, Sarah Moore
            </h5>
          </div>
        </article>
        <article
          class="sc-bcXHqe UJFxh"
        >
          <img
            alt="What the Ermine Saw"
            class="sc-gswNZR bUwCaT"
            src="https://books.google.com/books/publisher/content/images/frontcover/d_E3EAAAQBAJ?fife=w270"
          />
          <div
            class="sc-dkrFOg hXwiuy"
          >
            <h5
              class="sc-eDvSVe broPkG"
            >
              N/A
            </h5>
            <h3
              class="sc-hLBbgP degNmy"
            >
              What the Ermine Saw
            </h3>
            <h5
              class="sc-jSUZER hEWtoK"
            >
              Eden Collinsworth
            </h5>
          </div>
        </article>
        <article
          class="sc-bcXHqe UJFxh"
        >
          <img
            alt="What Artists Wear"
            class="sc-gswNZR bUwCaT"
            src="https://books.google.com/books/publisher/content/images/frontcover/gwJLEAAAQBAJ?fife=w270"
          />
          <div
            class="sc-dkrFOg hXwiuy"
          >
            <h5
              class="sc-eDvSVe broPkG"
            >
              N/A
            </h5>
            <h3
              class="sc-hLBbgP degNmy"
            >
              What Artists Wear
            </h3>
            <h5
              class="sc-jSUZER hEWtoK"
            >
              Charlie Porter
            </h5>
          </div>
        </article>
      </section>
      
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`BooksList should load more books when button clicked 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-eDWCr editcY"
      >
        <section
          class="sc-gKPRtg ilzhkz"
        >
          <article
            class="sc-bcXHqe UJFxh"
          >
            <img
              alt="Start Painting Now"
              class="sc-gswNZR bUwCaT"
              src="https://books.google.com/books/publisher/content/images/frontcover/osZVEAAAQBAJ?fife=w270"
            />
            <div
              class="sc-dkrFOg hXwiuy"
            >
              <h5
                class="sc-eDvSVe broPkG"
              >
                N/A
              </h5>
              <h3
                class="sc-hLBbgP degNmy"
              >
                Start Painting Now
              </h3>
              <h5
                class="sc-jSUZER hEWtoK"
              >
                Emily Powell, Sarah Moore
              </h5>
            </div>
          </article>
          <article
            class="sc-bcXHqe UJFxh"
          >
            <img
              alt="What the Ermine Saw"
              class="sc-gswNZR bUwCaT"
              src="https://books.google.com/books/publisher/content/images/frontcover/d_E3EAAAQBAJ?fife=w270"
            />
            <div
              class="sc-dkrFOg hXwiuy"
            >
              <h5
                class="sc-eDvSVe broPkG"
              >
                N/A
              </h5>
              <h3
                class="sc-hLBbgP degNmy"
              >
                What the Ermine Saw
              </h3>
              <h5
                class="sc-jSUZER hEWtoK"
              >
                Eden Collinsworth
              </h5>
            </div>
          </article>
          <article
            class="sc-bcXHqe UJFxh"
          >
            <img
              alt="What Artists Wear"
              class="sc-gswNZR bUwCaT"
              src="https://books.google.com/books/publisher/content/images/frontcover/gwJLEAAAQBAJ?fife=w270"
            />
            <div
              class="sc-dkrFOg hXwiuy"
            >
              <h5
                class="sc-eDvSVe broPkG"
              >
                N/A
              </h5>
              <h3
                class="sc-hLBbgP degNmy"
              >
                What Artists Wear
              </h3>
              <h5
                class="sc-jSUZER hEWtoK"
              >
                Charlie Porter
              </h5>
            </div>
          </article>
        </section>
        <button
          class="sc-jrcTuL dyKfxA"
        >
          Load more
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-eDWCr editcY"
    >
      <section
        class="sc-gKPRtg ilzhkz"
      >
        <article
          class="sc-bcXHqe UJFxh"
        >
          <img
            alt="Start Painting Now"
            class="sc-gswNZR bUwCaT"
            src="https://books.google.com/books/publisher/content/images/frontcover/osZVEAAAQBAJ?fife=w270"
          />
          <div
            class="sc-dkrFOg hXwiuy"
          >
            <h5
              class="sc-eDvSVe broPkG"
            >
              N/A
            </h5>
            <h3
              class="sc-hLBbgP degNmy"
            >
              Start Painting Now
            </h3>
            <h5
              class="sc-jSUZER hEWtoK"
            >
              Emily Powell, Sarah Moore
            </h5>
          </div>
        </article>
        <article
          class="sc-bcXHqe UJFxh"
        >
          <img
            alt="What the Ermine Saw"
            class="sc-gswNZR bUwCaT"
            src="https://books.google.com/books/publisher/content/images/frontcover/d_E3EAAAQBAJ?fife=w270"
          />
          <div
            class="sc-dkrFOg hXwiuy"
          >
            <h5
              class="sc-eDvSVe broPkG"
            >
              N/A
            </h5>
            <h3
              class="sc-hLBbgP degNmy"
            >
              What the Ermine Saw
            </h3>
            <h5
              class="sc-jSUZER hEWtoK"
            >
              Eden Collinsworth
            </h5>
          </div>
        </article>
        <article
          class="sc-bcXHqe UJFxh"
        >
          <img
            alt="What Artists Wear"
            class="sc-gswNZR bUwCaT"
            src="https://books.google.com/books/publisher/content/images/frontcover/gwJLEAAAQBAJ?fife=w270"
          />
          <div
            class="sc-dkrFOg hXwiuy"
          >
            <h5
              class="sc-eDvSVe broPkG"
            >
              N/A
            </h5>
            <h3
              class="sc-hLBbgP degNmy"
            >
              What Artists Wear
            </h3>
            <h5
              class="sc-jSUZER hEWtoK"
            >
              Charlie Porter
            </h5>
          </div>
        </article>
      </section>
      <button
        class="sc-jrcTuL dyKfxA"
      >
        Load more
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
